[DependencyInjection] Add "when" argument to #[AsAlias]

Zuruuh opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #60118
| License       | MIT

Also I wonder if it would make sense to accept an array of environments in `#[AsAlias]`, since in practice if I want to use the same service in two envs I'd have to duplicate the whole declaration now
```php
#[AsAlias(MyInterface::class, when: 'dev']
#[AsAlias(MyInterface::class, when: 'test']
class MyClass {}
```

Thoughts ?

Zuruuh committed:
[DependencyInjection] Add "when" argument to #[AsAlias]

nicolas-grekas commented:
Accepting string|array would make sense to me indeed.
Note that tests fail currently.

Zuruuh commented:
:+1: Updated the implem to use a `string|array` instead :+1: Also moved the check for existing aliases inside of the env condition to ensure we don't fail when an alias isn't going to be used for the selected environment.
Also fixed the huge diff on CHANGELOG.md I forgot to disable my auto-formatter 🤠 

The new tests pass now locally

alexandre-daubois commented on src/Symfony/Component/DependencyInjection/Attribute/AsAlias.php:
```diff
@@ -20,12 +20,20 @@
         public bool $public = false,
+        string|array $when = [],
     ) {
+        $this->when = is_array($when) ? $when : [$when];
```
You can simplify this part a bit:

```suggestion
        $this->when = (array) $when;
```

See https://3v4l.org/qlj0A

OskarStark commented on src/Symfony/Component/DependencyInjection/Tests/Loader/FileLoaderTest.php:
```diff
@@ -174,7 +177,8 @@ public function testRegisterClassesWithExcludeAsArray()
             new Definition(),
             'Symfony\Component\DependencyInjection\Tests\Fixtures\Prototype\\',
-            'Prototype/*', [
+            'Prototype/*',
+            [
```
please revert

Zuruuh replied:
Done :+1: Seems like I'm using a newer version of php-cs-fixer

nicolas-grekas commented:
Thank you @Zuruuh.
